datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -663,12 +914,32 @@ impl std::fmt::Debug for AggregationState {
 }
 
 impl GroupedHashAggregateStream {
+    /// Prune the groups from the `self.aggr_state.group_states` which are in
+    /// `GroupStatus::Emitted`(this status means that result of this group emitted/outputted already, and
+    /// we are sure that these groups cannot receive new rows.) status.
+    fn prune(&mut self) {
```

Comment:
I think this is good to keep the global `group_states` in a relatively small size. 
But it is possible that in some cases only a small percentage is pruned, and we pay for more 
copies.   I'm not sure when the prune will be triggered.

Reply:
`ordered_columns` store the section of the group by expression that defines ordering in `GroupState`. When a different `ordered_columns` is received, we are sure that previous groups with different `ordered_columns` are finalized (They will no longer receive new value). At the end of `group_aggregate_batch` we iterate over `self.aggr_state.group_states` and mark the groups that have different `ordered_columns` with the `ordered_columns` of the most recent (last) group as prunable.

As an example, If the table is like below, and we know that it satisfies `ORDER BY a ASC`
|a|
|----|
|1|
|1|
|2|
|2|
|3|
|3|

and group by clause is `GROUP BY a` group with `ordered_columns= Some(vec![1])` and `ordered_columns= Some(vec![2])` will be pruned. Since they are different than `ordered_columns= Some(vec![3])`. However, last group is not pruned because we still can receive values with 3 for column a